/*
 * Copyright 2018 The Feast Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package feast.serving;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";
import "feast/types/Value.proto";
import "feast/types/FeatureRow.proto";

option java_package = "feast.serving";
option java_outer_classname = "ServingAPIProto";
option go_package = "github.com/gojek/feast/sdk/go/protos/feast/serving";

service ServingService {
    // Get version information about this Feast serving.
    rpc GetFeastServingVersion (GetFeastServingVersionRequest) returns (GetFeastServingVersionResponse);

    // Get Feast serving store type: online or batch.
    rpc GetFeastServingType (GetFeastServingTypeRequest) returns (GetFeastServingTypeResponse);

    // Get online features synchronously.
    rpc GetOnlineFeatures (GetFeaturesRequest) returns (GetOnlineFeaturesResponse);

    // Get batch features asynchronously. 
    // 
    // The client should check the status of the returned job periodically by
    // calling ReloadJob to determine if the job has completed successfully 
    // or with an error. If the job completes successfully i.e.
    // status = JOB_STATUS_DONE with no error, then the client can check
    // the file_uris for the location to download feature values data.
    // The client is assumed to have access to these file URIs.
    rpc GetBatchFeatures (GetFeaturesRequest) returns (GetBatchFeaturesResponse);

    // Reload the job status with the latest state.
    rpc ReloadJob(ReloadJobRequest) returns (ReloadJobResponse);
}

message GetFeastServingVersionRequest {}

message GetFeastServingVersionResponse {
    string version = 1;
}

message GetFeastServingTypeRequest {}

message GetFeastServingTypeResponse {
    FeastServingType type = 1;
}

message GetFeaturesRequest {
    // List of feature sets and their features that are being retrieved
    repeated FeatureSet feature_sets = 1;

    // Dataset containing timestamp and entity id data. Used during retrieval of feature rows
    // and for joining feature rows into a final dataset
    EntityDataset entity_dataset = 2;

    message FeatureSet {
        // Feature set name
        string name = 1;

        // Feature set version
        int32 version = 2;

        // Features that should be retrieved from this feature set
        repeated string feature_names = 3;

        // The features will be retrieved if:
        // entity_timestamp - max_age  <= event_timestamp <= entity_timestamp
        //
        // If unspecified the default max_age specified in FeatureSetSpec will 
        // be used.
        google.protobuf.Duration max_age = 4;
    }

    message EntityDataset {
        // List of entity names contained within this incoming request. Each entity name is globally
        // unique within Feast. The user is assumed to have used the exact column name in their
        // EntityDataset if they are providing this dataset through a batch process.
        repeated string entity_names = 1;

        // List of Unix epoch entity_timestamp and entity_id values
        repeated EntityDatasetRow entity_dataset_rows = 2;
    }

    // EntityDatasetRow specifies:
    // - the timestamp range over which feature values should be retrieved (required for batch serving)
    // - the specific entity ids that should be retrieved (required for online serving)
    //
    // If there are duplicate entity ids for the same timestamp range, only the 
    // one with the latest event_timestamp will be retrieved.
    //
    // Entity ids may be ommitted for batch features retrieval. In this case,
    // all entities with distinct entity ids within the valid timestamp range 
    // will be retrieved.
    message EntityDatasetRow  {
        // entity_timestamp is the upper bound of the timestamp range over 
        // which the feature values should be retrieved.
        //
        // For online serving entity_timestamp is optional (ignored), as the 
        // latest is always retrieved.
        // 
        // The timestamp range is defined as follows:
        // entity_timestamp - max_age <= event_timestamp <= entity_timestamp
        google.protobuf.Timestamp entity_timestamp = 1;

        // The entity ids for which the feature values should be retrieved.
        // 
        // The order of the values should follow that in entity_names in EntityDataset.
        // For online serving, it is required to specify entity_ids.
        // For batch serving, it is optional.
        repeated feast.types.Value entity_ids = 2;
    }
}

message GetOnlineFeaturesResponse {
    // A FeatureDataSet is returned for each feature set in the incoming request
    repeated FeatureDataset feature_datasets = 2;

    // The FeatureDataSet contains information about the Feature Set in the incoming request,
    // as well as feature data that can be joined to the incoming EntityDataSet. The row count
    // for the returning FeatureDataSet will match that of the row count for the incoming
    // EntityDataSet.
    // If any of the keys do not have values, empty feature rows will be returned.
    message FeatureDataset {
        // Feature set name
        string name = 1;

        // Feature set version
        int32 version = 2;

        // Each feature data set contains a list of feature rows. The timestamps within the row
        // are the original event timestamps from when that row was written to the backing store.
        // When these FeatureRows are joined to the EntityDataSetRows, the FeatureRow timestamps
        // will be dropped in favour of the EntityDataSetRow timestamp.
        repeated feast.types.FeatureRow feature_rows = 3;
    }
}

message GetBatchFeaturesResponse {
    Job job = 1;
}

message ReloadJobRequest {
    Job job = 1;
}

message ReloadJobResponse {
    Job job = 1;
}

enum FeastServingType {
    FEAST_SERVING_TYPE_INVALID = 0;
    // Online serving receives entity data directly and synchronously and will 
    // respond immediately.
    FEAST_SERVING_TYPE_ONLINE = 1;
    // Batch serving receives entity data asynchronously and orchestrates the 
    // retrieval through a staging location.
    FEAST_SERVING_TYPE_BATCH = 2;
}

enum JobType {
    JOB_TYPE_INVALID = 0;
    JOB_TYPE_DOWNLOAD = 1;
}

enum JobStatus {
    JOB_STATUS_INVALID = 0;
    JOB_STATUS_PENDING = 1;
    JOB_STATUS_RUNNING = 2;
    JOB_STATUS_DONE = 3;
}

enum DataFormat {
    DATA_FORMAT_INVALID = 0;
    DATA_FORMAT_CSV = 1;
    DATA_FORMAT_PARQUET = 2;
    DATA_FORMAT_AVRO = 3;
    DATA_FORMAT_JSON = 4;
}

message Job {
    string id = 1;
    // Output only. The type of the job.
    JobType type = 2;
    // Output only. Current state of the job.
    JobStatus status = 3;
    // Output only. If not empty, the job has failed with this error message.
    string error = 4;
    // Output only. The list of URIs for the files to be downloaded or 
    // uploaded (depends on the job type) for this particular job.
    repeated string file_uris = 5;
    // Output only. The data format for all the files.
    // For CSV format, the files contain both feature values and a column header.
    DataFormat data_format = 6;
}
